### PR TITLE
Refactor CBPS functions as classes and add an example +  package it

### PR DESCRIPTION
**Pull request by ivanhigueram**
Main changes in this PR: 
 - Refactor `cbps_torch.py` as a class with the same functionality as the original.
 - Test the class using the `lalonde.R` dataset from [LalRUtils](https://github.com/apoorvalal/LalRUtils/blob/master/data/lalonde.psid.RData).
 - Create `project.toml` for future packaging and testing
 - Add some small docstrings just to guide myself. 

Some changes in the `CBPS` implementation:
 - Add L1-norm regularization to loss during optimization. Not sure if that's a desirable feature, but is there.

### CODE REVIEW

File: src/cbps_torch.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        """Estimate weights optimizing loss function."""
+        optimizer = torch.optim.Adam(self.theta, lr=self.lr)
+
+        # Iteratively and update theta vector
+        for iteration in range(self.niter):
+            optimizer.zero_grad()
+            loss = self.loss_function(self.theta, self.X_t, self.w_t)
+
+            if self.reg:
+                loss += self.reg * torch.linalg.vector_norm(self.theta, 1)
```
Comment (apoorvalal) [inline src/cbps_torch.py:141]: nice! why prefer L1 over L2 though? Do you want sparsity?

File: src/cbps_retarget.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        loss = torch.mean(torch.exp(torch.matmul(X0, theta))) - torch.matmul(
+            X1, theta
+        )  # target moments
+        return loss
+
+    optimizer = torch.optim.Adam([theta], lr=lr)
+    # Iteratively apply gradient descent to minimize the loss function
+    for iteration in range(niter):
+        optimizer.zero_grad()  # Clear previous gradients
+        loss = loss_function(theta, X0, X1)  # Calculate loss
```
Comment (apoorvalal) [inline src/cbps_retarget.py:52]: for parity with disaggregated data case, also add regularization here. These are implicitly the same program except we only feed in summary statistics in the retarget version

File: src/cbps_torch.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+        return None
+
+    def parameters(self):
+        """Return final parameters after optimization."""
+        if self.estimand == "ATT":
+            out = np.exp(self.X_t @ self.theta)[self.w_t == 0]
+        elif self.estimand == "ATE":
+            out = np.exp(self.X @ self.theta)
+
```
Comment (apoorvalal) [inline src/cbps_torch.py:166]: meant to do this but now going to give you hw from 450b: add a diagnose() method that visualizes pre- and post-reweighting balance along covariates. Something like this https://cran.r-project.org/web/packages/cobalt/vignettes/love.plot.html

File: src/cbps_torch.py [hunk 4]
```diff
@@ -19,25 +19,65 @@ def __init__(
...
         if self.estimand not in ["ATT", "ATE"]:
             raise NotImplementedError("Estimand in ['ATT', 'ATE'] supported")
 
         # Define parameters for calculation
         self.n, self.p = X.shape
-
-        self.X = torch.from_numpy(X).float().to(device)
         self.W = torch.from_numpy(W).float().to(device)
 
+        # SVD for dimensionality reduction
```
Comment (apoorvalal) [inline src/cbps_torch.py:65]: add cite to https://www.jmlr.org/papers/volume19/17-777/17-777.pdf once you test this out. Have never thought low rank is all that useful in generic balancing problems, but makes sense for panel applications in particular. If you do this, allow tuning of k (`self.svd`).

**Review by apoorvalal — APPROVED** [review]

**Comment by apoorvalal** [comment]
LGTM. Thanks!

File: src/cbps_torch.py (hunk 1 above)
Comment (ivanhigueram) [reply src/cbps_torch.py:141]: Yeah, I thought sparsity would be a good goal, but we can keep it flexible and add L2 too.

File: src/cbps_retarget.py (hunk 2 above)
Comment (ivanhigueram) [reply src/cbps_retarget.py:52]: Huh, i was confused how to integrate those `X1`, can you expand what `retargeting` means here?

Comment (apoorvalal) [reply src/cbps_retarget.py:52]: `retargeting` here means you're estimating balancing weights that calibrate a sample `X0` towards a population `X1`, where you only have summary statistics for the target population `X1`. This is useful in a bunch of settings where external validity is something you try to achieve rather than handwave :laughing:.  
In the ATT estimation case, you have individual level data for everyone, but the loss function only wants to match the first moments between the `X0` mean and `X1` mean, so the individual level data for the treated units wasn't actually used. So the two functions `_torch` and `_retarget` are basically the same except in the first one, averaging over the treatment group is [implicit](https://github.com/apoorvalal/covariate_balancing_propensity_scores/blob/master/src/cbps_torch.py#L106) and in the second one, the averaging is explicit since the [loss function accepts the target averages directly](https://github.com/apoorvalal/covariate_balancing_propensity_scores/blob/master/src/cbps_retarget.py#L43). Check out the Wang/Zubizarreta paper linked in the readme, it is really good.
